### DISCUSSION TITLE
VTM MultiMapDatabase deduplication like Mapsforge MultiMapDataStore

### DISCUSSION THREAD

**Discussion by demiantres** [mapsforge/vtm#897] [Ideas]
Hi,

The VTM MultiMapDatabase generally works fine but sometimes there are problems with areas defined in different map files (it looks as if the areas are completely removed/clipped away).

Does the VTM implementation include deduplication mechanisms, similar to the old Mapsforge MultiMapDataStore.DataPolicy (https://github.com/mapsforge/mapsforge/blob/master/mapsforge-map/src/main/java/org/mapsforge/map/datastore/MultiMapDataStore.java)? The MultiMapDatabase contains methods with a deduplication parameter (for instance MultiMapDatabase .readLabels) but I see no way how to use them.

Regards,

**Comment by devemux86** [comment]
About overlapping polygons in OpenGL, see https://github.com/mapsforge/vtm/issues/224 and https://github.com/mapsforge/vtm/issues/806#issuecomment-790602712.

I have not transferred Mapsforge's multi map data policy in VTM (see https://github.com/mapsforge/vtm/issues/208),
as VTM works differently with `ITileDataSink` instead of `MapReadResult`.

Mapsforge's `RETURN_FIRST` has empty areas and `DEDUPLICATE` was slow.

_"[DEDUPLICATE](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-map/src/main/java/org/mapsforge/map/datastore/MultiMapDataStore.java#L33-L35): the data from all files will be returned but duplicates will be eliminated. This is suitable when multiple maps cover the different areas, but there is some overlap at boundaries. This is the most expensive operation and often it is actually faster to double paint objects."_

**Reply by demiantres** [reply]
The patch from #806 works pretty well.

**Reply by devemux86** [reply]
> The patch from #806 works pretty well.

It also has empty areas.

**Reply by demiantres** [reply]
Good to know. But the overall look of the maps is better than without the patch.

**Reply by devemux86** [reply]
If you can make it work properly,
I can include it via a parameter in `MultiMapFileTileSource`.

It may need a larger data type for the hash (long, double, ...)
and probably adding more `MapElement` fields in the hash.

**Comment by devemux86** [comment]
https://github.com/mapsforge/mapsforge/pull/1288 optimizes the deduplication of multiple maps in Mapsforge.

**Comment by devemux86** [comment]
#903 #905 provide in VTM:
- Mapsforge: deduplicate maps
- Fix overlapping map regions
